File: tests/js/src/inngest/sdk_parallel_fan_in_test.ts
import { inngest } from "@/inngest/client";

// Helper function invoked by testParallelFanIn.  Sleeps a random 1-5s so
// concurrent invokes complete at different times, widening the race window.
export const sleepRandom = inngest.createFunction(
  { id: "sleep-random" },
  { event: "tests/sleep-random.test" },
  async ({ step }) => {
    const ms = 1000 + Math.floor(Math.random() * 4000);
    await step.sleep("zzz", `${ms}ms`);
    return { slept: ms };
  }
);

// Regression test for parallel fan-in runs hanging.  Drives 40 concurrent
// ops (20 step.run, 20 step.invoke) through one Promise.all and asserts the
// parent run completes with the expected aggregate.  step.run results sum to
// a known constant; the invoke count confirms every child finished — so the
// Go driver catches both "run stalled" and "step ran the wrong number of
// times" regressions.
export const testParallelFanIn = inngest.createFunction(
  { id: "parallel-fan-in" },
  { event: "tests/parallel-fan-in.test" },
  async ({ step }) => {
    const tasks: Promise<unknown>[] = [];

    for (let i = 0; i < 20; i++) {
      tasks.push(
        step.invoke(`invoke-${i}`, {
          function: sleepRandom,
          data: { name: "tests/sleep-random.test", data: {} },
        })
      );
    }

    for (let i = 0; i < 20; i++) {
      tasks.push(step.run(`run-${i}`, () => i * i));
    }

    const results = await Promise.all(tasks);

    // First 20 results are invokes (random sleep duration); last 20 are
    // step.run squares.  Sum the squares — 0² + 1² + … + 19² = 2470 — and
    // count completed invokes by checking the response shape.
    const invokeResults = results.slice(0, 20);
    const stepResults = results.slice(20) as number[];

    const stepSquares = stepResults.reduce((a, b) => a + b, 0);
    const invokeCount = invokeResults.filter(
      (r: any) => r && typeof r.slept === "number"
    ).length;

    return { stepSquares, invokeCount };
  }
);

File: tests/sdk_parallel_fan_in_test.go
package main

import (
	"context"
	"encoding/json"
	"testing"
	"time"

	"github.com/inngest/inngest/tests/client"
	"github.com/inngest/inngestgo"
	"github.com/stretchr/testify/require"
)

// TestParallelFanIn is a regression test for parallel fan-in runs hanging.
// It drives the JS testParallelFanIn function (40 concurrent ops in one
// Promise.all: 20 step.run + 20 step.invoke of sleepRandom) and asserts the
// parent run completes with the expected aggregate.  If discovery dedup ever
// decouples from pending-step tracking again, the run stalls and this test
// fails on the wait-for-status timeout.  If a step runs the wrong number of
// times the aggregate assertions catch it.
func TestParallelFanIn(t *testing.T) {
	registerDirect(t)
	defer deregisterDirect(t)

	cli := client.New(t)
	ctx := context.Background()
	start := time.Now().Add(-2 * time.Second)

	eventURLStr := eventURL.String()
	ic, err := inngestgo.NewClient(inngestgo.ClientOpts{
		AppID:    "test",
		EventKey: &eventKey,
		EventURL: &eventURLStr,
	})
	require.NoError(t, err)
	_, err = ic.Send(ctx, inngestgo.Event{
		Name: "tests/parallel-fan-in.test",
		Data: map[string]any{},
	})
	require.NoError(t, err)

	runID := waitForRecentRun(t, cli, ctx, "parallel-fan-in", start, "COMPLETED", 60*time.Second)
	run := cli.Run(ctx, runID)
	require.Equal(t, "COMPLETED", run.Status)

	var out struct {
		StepSquares int `json:"stepSquares"`
		InvokeCount int `json:"invokeCount"`
	}
	require.NoError(t, json.Unmarshal([]byte(run.Output), &out), "run output: %s", run.Output)

	// 0² + 1² + ... + 19² = 2470
	require.Equal(t, 2470, out.StepSquares, "sum of step.run squares — wrong value means a step ran the wrong number of times or returned the wrong value")
	require.Equal(t, 20, out.InvokeCount, "every step.invoke must return a {slept: number} response")
}
